### PR TITLE
[AutoDiff][Do not merge] Complete primal and adjoint synthesis for single BB automatic differentiation

### PR DESCRIPTION
**Pull request by rxwei**
Starting an initial PR for code review. In the meantime I'll try to split this to smaller PRs for more targeted review.

Some main components are:

0. `DifferentiableActivityAnalysis`
1. `PrimalGen` and `PrimalGenCloner`
2. `AdjointGen` and `AdjointEmitter`
3. Symbolic `AdjointValue` data structure
4. "Delayed gradient" support.

### CODE REVIEW

File: include/swift/Syntax/TokenKinds.def [hunk 1]
```diff
@@ -23,6 +23,7 @@
 ///     SIL_KEYWORD(kw)
 ///   POUND_KEYWORD(kw)
 ///     POUND_OBJECT_LITERAL(kw, desc, proto)
+///     POUND_OLD_OBJECT_LITERAL(kw, new_kw, old_arg, new_arg)
```
Comment (dan-zheng) [inline include/swift/Syntax/TokenKinds.def]: This code was removed in https://github.com/apple/swift/commit/bdcd87754d181fcb26b23276b8a82129e027f8b9. Same for the other changes in this file.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 2]
```diff
@@ -52,86 +60,420 @@
...
+  void print(llvm::raw_ostream &os) const;
 };
-} // end anonymous namespace
 
-namespace llvm {
-using swift::SILFunction;
-using swift::SILReverseDifferentiableAttr;
+/// Information about the primal function produced by PrimalGen, e.g.
+/// mappings from the original values to their corresponding ones in the primal
+/// value struct produced by the primal function.
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: By "primal value struct", you must be referring to the struct containing primal intermediate values.
An explanation of the purpose of the primal value struct would be nice, e.g. "the primal value struct stores intermediate values from the original function".

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 3]
```diff
@@ -405,342 +861,1506 @@ bool ADContext::supportsVectorDifferentiation(Type type) const {
...
+  auto *scalarTyDecl = scalarTy.getAnyNominal();
+  assert(scalarTyDecl && "ScalarElement must be a nominal type");
+  // %0 = ... : $<scalar type>
+  auto scalarBuf =
+    builder.createAllocStack(loc, SILType::getPrimitiveObjectType(scalarTy));
+  convertIntToIndirectExpressible(value, scalarTyDecl, scalarBuf,
+                                  loc, builder, context);
+  auto scalarLOQ = getBufferLOQ(scalarTy, module);
+  auto scalarVal = builder.createLoad(loc, scalarBuf, scalarLOQ);
+  // dealloc_stacl %0 : $*<scalar type>
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Typo: `dealloc_stacl` -> `dealloc_stack`

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 4]
```diff
@@ -405,342 +861,1506 @@ bool ADContext::supportsVectorDifferentiation(Type type) const {
...
+};
+} // end anonymous namespace
+
+/// Creates a struct declaration (without contents) for storing contextual
+/// information for a function. The newly created struct will have the same
+/// generic parameters as the function.
+///
+/// This helper function will be used to create checkpointing structs,
+/// specifically structs for control-independent checkpoints and tape groups.
+StructDecl *ADContext::createDependentStructForFunction(SILFunction *function) {
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Is there a reason for the difference in terminology between "dependent struct" and "primal value struct"?

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 5]
```diff
@@ -756,384 +2376,990 @@ AdjointGen::createAdjointFunction(SILFunction *original,
...
+    assert(all_of(llvm::zip(type->getElementTypes(), elements),
+      [](std::pair<Type, AdjointValue> pair) {
+        return pair.first->isEqual(
+          pair.second.getType().getASTType());
+      }));
+    return getAggregate(Kind::Tuple, silTy, elements, allocator);
+  }
+  
+  ArrayRef<AdjointValue> getTupleElements() const {
+    assert(kind == Kind::Tuple);
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Use `assert(isTuple());`?

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 6]
```diff
@@ -1402,9 +3612,54 @@ class Differentiation : public SILModuleTransform {
...
+  else {
+    if (auto loc = gi->getLoc()) {
+      auto *expr = loc.castToASTNode<ReverseAutoDiffExpr>();
+      context.diagnose(expr->getOriginalExpr()->getLoc(),
+                       diag::autodiff_opaque_function_unsupported);
+    }
+    context.setErrorOccurred();
+    return;
+  }
+  // We invalide analyses on the parent function because the `gradient`
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Typos: `invalide` -> `invalidate`, on the next line `instruciton` -> `instruction`.

File: test/AutoDiff/autodiff_diagnostics.swift [hunk 7]
```diff
@@ -2,4 +2,38 @@
 
 func foo(_ f: @escaping (Float) -> Float) -> Float {
   return #gradient(f)(0) // expected-error {{differentiating an opaque function is not supported yet}}
+
+// func foo(_ f: @escaping (Float) -> Float) -> Float {
```
Comment (dan-zheng) [inline test/AutoDiff/autodiff_diagnostics.swift]: Can this commented testcase be removed?

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp (hunk 5 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp (hunk 6 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Done.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp (hunk 2 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Added explanation.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp (hunk 3 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Done.

File: include/swift/Syntax/TokenKinds.def (hunk 1 above)
Comment (rxwei) [reply include/swift/Syntax/TokenKinds.def]: This got in accidentally during rebasing.

**Comment by rxwei** [comment]
Splitting this into a few functional pieces tonight:
1. Add `SILGradientOptions` and support computing type signatures for delayed gradients.
2. Implement `DifferentiationActivityAnalysis`.
3. Enhance `DifferentiationTask`.
4. Add `PrimalInfo`.
5. Implement primal generation.
6. Add symbolic `AdjointValue` for efficiently differentiating aggregate data structures including tuples and structs.
7. Implement adjoint generation.

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 8]
```diff
@@ -52,86 +60,546 @@
...
+  ///     %prim_vals = tuple (%prim_val_0, ..., %prim_val_n)     [CHECKPOINT]
+  ///     %orig_results = tuple (%orig_res_0, ..., %orig_res_n)  [CHECKPOINT]
+  ///
+  /// If this function is non-control-dependent, primal values will be
+  /// checkpointed into the primal value struct as a tuple member, and get
+  /// inserted into `nestedStaticPrimalValueMap`. Otherwise, it'll go to the
+  /// corresponding tape of its type.
+  DenseMap<ApplyInst *, VarDecl *> nestedStaticPrimalValueMap;
+  
+  /// Mapping from `apply` instructions to their
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Unfinished doc comment.

**Pull request closed without merging**